docs/Configuration.md
# Configuration

*JSCAS* uses [nixconfig](https://npm.im/nixconfig),
[nixconfig-toml](https://npm.im/nixconfig-toml), and
[nixconfig-yaml](https://npm.im/nixconfig-yaml) for loading configuration.
This nixconfig name is `jscas`. Thus, a `jscas.js`, `jscas.json`, `jscas.toml`,
`jscas.yaml`, or `jscas.yml` file within any of the following directories will
be loaded:

+ `/etc`
+ `/etc/jscas`
+ `~`
+ `~/.jscas`
+ `~/.config/jscas`

It is also possible to set the environment variable `nixconfig_config_home`
to specify any other directory. This directory will be searched last, and the
configuration within it will take precedence.

<a id="default"></a>
## Default Configuration

The default configuration is viewable by reading
[/lib/config.js](/lib/config.js). The default configuration attempts to have
"sane" values. There are some values that *must* be adjusted for any install
to function properly, but a great deal of effort is made to keep this to
a minimum.

In particular, the server, and protocol itself, relies on cookies to function.
As a result, the most important configuration parameters deal with:

1. The cookie settings for the application to keep track of session data.
1. The cookie settings for the ticket granting ticket.

<a id="parameters"></a>
## Parameters

Each heading in this section is a property on a root configuration object.

<a id="logger"></a>
### `logger`

The `logger` configuration property configures the included
[Pino](https://npm.im/pino) instance. To learn about the available options,
read
[Pino's documentation](https://github.com/pinojs/pino/blob/49cb060f3f349bc90/docs/API.md#constructor).

<a id="server"></a>
### `server`

An object with the properties:

+ `address` (Default: `'127.0.0.1'`): the local IP address to bind to.
+ `port` (Default: `9000`): the local port to use for listening.

<a id="caching"></a>
### `caching`

*JSCAS* uses an internal cache for things such as session storage. It is
**important** that this configuration be updated to supply a proper backing
store. The default uses an in-memory store that should not be relied upon
in production.

This parameter accepts an object with the following properties:

+ `privacy` (Default: `'no-cache'`): any value that is valid for the
`Cache-control` header.
+ `cacheSegment` (Default: `'jscas'`): the namespace to store cache values
under.
+ `cache` (Default: an in-memory cache): This is an
[abstract-cache](https://www.npmjs.com/package/abstract-cache) configuration
object. An alternate `driver` should be specified to prevent the usage
of the in-memory storage.

<a id="helmet"></a>
### `helmet`

An configuration object to pass to [fastify-helmet][fastify-helmet]. The
default value is `{}`.

[fastify-helmet]: https://github.com/fastify/fastify-helmet

<a id="data-sources"></a>
### `dataSources`

*JSCAS* makes various data sources available to registered plugins. The plugins
to be used will dictate this configuration parameter. The bundled demonstration
plugins do not require any data sources to be defined.

This parameter accepts an object with the following properties:

+ `mongodb`: configuration for the standard
[MongoDB client](http://mongodb.github.io/node-mongodb-native/3.0/reference/connecting/).
+ `postgres`: configuration for the standard
[pg module](https://node-postgres.com/features/connecting).
+ `redis`: configuration for the
[ioredis module](https://www.npmjs.com/package/ioredis#connect-to-redis).

<a id="cookie"></a>
### `cookie`

This parameter defines the settings that will be used when server sets
ticket granting ticket cookies. It is **important** this parameter be configured
properly.

This parameter accepts an object with the following properties:

+ `domain` (Default: `'.cas.example.com'`): this should be set to the *specific
domain* for the CAS server.
+ `path` (Default: `'/'`): this should be set to the base path for the CAS
server.
+ `expires` (Default: `1800000`): time, in milliseconds, that the cookie should
be valid. This time should be less than or equal to the cache lifetime.
+ `secure` (Default: `true`): specifies if the cookie should only be sent over
HTTPS connections.
+ `sameSite` (Default: `true`): dictates if the cookie is allowed to be sent
with cross site requests. See
[https://www.owasp.org/index.php/SameSite](https://www.owasp.org/index.php/SameSite).
+ `httpOnly` (Default: `true`): dictates if the cookie can be read by scripts
within a web page.

<a id="session"></a>
### `session`

This parameter defines the settings governing the session used by the server
to track users's authorization, and by plugins for various purposes.

The parameter accepts an object with the following properties:

+ `secretKey` (Default: **insecure**): it is **imperitive** this setting be
adjusted. It must be a string of at least 32 characters. This setting is used
to sign the session cookie.
+ `sessionMaxAge` (Default: `1800000`): time, in milliseconds, that the session
will be valid. This should be less than or equal to the
[cookie lifetime](#cookie) (typically equal).
+ `cookie`: the same possible settings as in [cookie](#cookie). It is
recommended that the values be the same as the [cookie](#cookie) configuration.

<a id="tickets"></a>
### `tickets`

This parameter defines settings for the various tickets that comprise the
CAS protocol.

This parameter accepts an object with the following properties:

+ `serviceTicket`: an object with the property `ttl` that defines the maximum
lifetime, in milliseconds, that a service ticket is valid. The default lifetime
is 10 seconds. This should be a very short time since a service ticket is only
valid for one service verification.
+ `ticketGrantingTicket`: an object with the properties `cookieName` and `ttl`
The `cookieName` defaults to `'jscas-tgt'`. The `ttl` defaults to `1800000`;
this is the time, in milliseconds, that the ticket granting ticket will be
valid. This should match the maximum age of the session.

<a id="v3overv2"></a>
### `v3overv2`

This parameter toggles the ability to serve CAS protocol version 3 attributes
to services via the CAS protocol version 2 endpoints. This parameter is a
boolean that defaults to `false`.

<a id="saml11Banner9Hack"></a>
### `saml11Banner9Hack`

This parameter toggles inclusion of a `UDC_IDENTIFIER` attribute in
`/samlValidate` responses. This is to enable compliance with the SIS product
Banner 9. The value of the attribute will be set to the same value as the
regular CAS `user` property, i.e. the "username".

This parameter is a boolean that defaults to `false`.

Note: the value of `UDC_IDENTFIER` gets mapped to a value in Banner's
`general.gobumap` table.

<a id="plugins"></a>
### `plugins`

This parameter defines the set of plugins that the server will use for various
functionalities. A basic set of plugins is required for the server to function
properly. Each value should be the name of a module that provides a
corresponding plugin. If the name begins with `~/`, then the plugin is loaded
from the server's `lib/plugins/` directory; this is only useful for the
bundled demonstration plugins.

Any plugin name containing `>` indicates a nested plugin. Such plugins are a
module that exports a hash of plugins. For example, if a plugin named `foo`
returns the object:

```js
{
  bar: function () {},
  baz: function () {}
}
```

Then the name `foo>bar` will load the `bar` plugin exported by the `foo`
module.

This parameter accepts an object with the following properties:

+ `attributesResolver` (Default: `~/jsAttributesResolver`): name of the module
that provides an attributes resolver plugin. This plugin is required to properly
support CAS protocol version 3.0, but is otherwise optional.
+ `theme` (Default: `~/basicTheme`): name of the module that provides the
user interface. This is a **required** plugin.
+ `serviceRegistry` (Default: `~/jsServiceRegistry`): name of the module that
provides the service registry. This is a **required** plugin.
+ `ticketRegistry` (Default: `~/jsTicketRegistry`): name of the module that
provides the ticket registry. This is a **required** plugin.
+ `auth` (Default: `['~/jsIdP']`): this is an array of plugins that provide
authenticators. At least one authenticator is **required**.
+ `misc` (Default: `[]`): this is an array of plugins that provide
miscellaneous additional functionality. No plugins are required in this
category.

<a id="plugins-conf"></a>
### `pluginsConf`

This parameter defines the configuration for each of the supplied
[plugins](#plugins).

This parameter accepts an object with properties that are equal to the names
of the respective plugin they are for. As an example, if there is a
miscellaneous plugin with the name "foo" registered, this parameter may
look like:

```js
{
  foo: {
    option1: 'value',
    option2: 'value'
  }
}
```


Changelog.md
### 0.15.1
+ See https://github.com/jscas/cas-server/releases for all future release notes.

### 0.15.0
+ Update dependencies.
+ Add OpBeat client to plugin phase1 init context.
+ Switch to `node-tap` for testing.

### 0.14.0
+ [breaking] Change the `server.cache` configuration property to `server.catbox`.

  This change allows us to define the configuration while retaining OpBeat
  instrumentation for catbox drivers like Redis or MongoDB.

### 0.13.1
+ Fix bug that was preventing modules from loading correctly.

### 0.13.0
+ [breaking] Changes Pino configuration option `pretty` to `prettyPrint`.

  Pino now natively supports enabling pretty printing via the `prettyPrint`
  option being passed to its constructor. So we now directly use it.
+ Adds support for OpBeat instrumentation.

  The [OpBeat][opbeat] client has been integrated into the server.
  The integration is minimal; it does not directly send any errors to the
  service. This may change in later version. See the notes in
  `settings.example.js` and the Readme for more information.

  [opbeat]: https://opbeat.com/

### 0.12.1
+ Fix `package.json` so we can actually publish.

### 0.12.0
+ Update dependencies.
+ Add support for `renewal` parameter to `/login`.
+ Update usage of deprecated Marko template methods.
+ Migrate request logging to `request.logger` instead of `request.log`.

  This change is meant to improve performance. Using `request.log` meant we
  were waiting on Hapi and Pino to process tags. Also, we couldn't use format
  strings, e.g. `log('foo %s', 'bar')`. This meant unnecessary creation of
  strings from template strings.
+ Add support for `service` parameter on `/logout`. The server will redirect
  you back to the service if it matches a service URL in the service registry.

### 0.11.0
+ Update dependencies.
+ Add `hapi-pino` for request logging.
+ Add `md/Logging.md` doc to establish logging standard and update code accordingly.
+ Add API stablity notice to readme.
+ Add `Promise` object to plugin Phase 1 context.

  The `Promise` object is an instance of the [Bluebird][bluebird] promise
  library. The [bluebird-co][bbco] library is also added to the object. This
  allows plugin developers to use the same promise library as the server, and
  get some free coroutine handlers.

  Bluebird was chosen as an easy performance improvement since it is
  [faster than native promises][faster-promises].
+ Switch to [cas-server-pg-registries][pg-registries] as the default service
  and ticket registries.

  Upon speaking with @mcollina at Node Interactive 2016 (North America), it
  was decided to drop the MongoDB registries due to the way [Mongoose][mongoose]
  does some things. Specifically, it is really slow at merging document references.

  We had originally switched to the MongoDB registries to improve performance.
  This was because the original PostgreSQL registries were very slow. The issue
  with those registries is that they rely on the [Objection][objection] ORM.
  A combination of the natural slowness of ORMs and misunderstanding of the
  way Objection works led to the poor performance.

  The new PostgreSQL registries do not use any ORM. They directly query the
  database through hand crafted SQL statements. These registires should not
  pose a performance risk.
+ `postgres` data source added.
+ `knex` as a supported data source has been removed. It has been replaced
  with the `postgres` data source.
+ `settings.example.js` updated to reflect `postgres`, `knex`, and new registries.
+ {breaking} Ticket registry specification updated to require `getSTbyTGT` to
  return an array of results instead of the first match.

[bluebird]: http://bluebirdjs.com/docs/getting-started.html
[bbco]: https://www.npmjs.com/package/bluebird-co
[faster-promises]: http://bluebirdjs.com/docs/benchmarks.html
[pg-registries]: https://github.com/jscas/cas-server-pg-registries
[mongoose]: https://www.npmjs.com/package/mongoose
[objection]: https://www.npmjs.com/package/objection

### 0.10.1
+ Support authentication plugins returning a boolean instead of a rejection.

  The `jscas-ad-auth` plugin returns `false` on failure instead of throwing an
  error. This caused the server to accept garbage usernames.

### 0.10.0
+ Load datasources synchronously because some plugins may depend on the
  connections being established during phase 1

+ Remove Marko from the phase 1 context because Marko templates may not work
  if the module is not in the same path of the plugin that uses it. This can
  happen if configuration is in its own project outside of the `cas-server`
  project (installation)

### 0.9.0
+ Migrate more functionality to `casInterface`

+ Update route loading to support Hapi 15

+ Update dependencies

+ Fix errors with expired state cookies in situations where the page should
  just load regardless of the old state

### 0.8.0
+ Update `preAuth` hook signature

  The `preAuth` hook signature was updated to accept a single parameter.
  The passed in parameter is an object with properties matching the original
  parameter list. However, the `loginTicket` property was removed and a
  `cas` object added (maps to the object defined in `lib/casInterface.js`).

  This change affects the `clearpass` plugin.

+ Removes support for login tickets

  The login ticket is an [optional part](https://github.com/apereo/cas/issues/1939)
  of the specification. Implementing them is too much of a burden and is more
  easily solved via a CSRF prevention token that doesn't require a ticket
  registry.

  This change affects theme and registry plugins.

+ New config block: `loginCSRF`

  The login form now uses a CSRF prevention token that isn't a login ticket.
  To support this, a new configuration block was added: `loginCSRF`. See
  `settings.example.js` for details.

### 0.7.0
+ Update dependencies
  + knex is no longer a hard dependency
+ Add a CAS API object and switch to using it for processing logins


lib/config.js
'use strict'

// This is the default configuration for the server. For a description of
// each parameter see the "Configuration.md" document in the `/docs`
// directory.

const path = require('path')
const isDebug = require('isdebug')
const nixconfig = require('nixconfig')
const fastifyCaching = require('fastify-caching')

const maxAge = 1800000 // 30 minutes in milliseconds
const cookieOptions = {
  domain: '.cas.example.com',
  path: '/',
  expires: maxAge,
  secure: true,
  // If testing with localhost sites and Chrome, use 'lax'
  sameSite: true,
  httpOnly: true
}

const initialConfig = {
  logger: {
    name: 'jscas-server',
    prettyPrint: isDebug,
    level: (isDebug) ? 'debug' : 'info'
  },

  server: {
    address: '127.0.0.1',
    port: 9000
  },

  caching: {
    // This should always be NOCACHE.
    privacy: fastifyCaching.privacy.NOCACHE,
    cacheSegment: 'jscas',
    cache: {
      driver: {
        name: undefined,
        options: {}
      }
    }
  },

  helmet: {},

  dataSources: undefined,
  // dataSources: {
  //   mongodb: {},
  //   postgres: {},
  //   redis: {}
  // }

  // Options for fastify-cookie.
  cookie: cookieOptions,

  session: {
    secretKey: 'some-secret-password-at-least-32-characters-long',
    sessionMaxAge: maxAge,
    cookie: cookieOptions
  },

  tickets: {
    serviceTicket: {
      ttl: 10000
    },
    ticketGrantingTicket: {
      cookieName: 'jscas-tgt',
      ttl: maxAge
    }
  },

  // Set to `true` to return user attributes via the `/serviceValidate` endpoint.
  v3overv2: false,

  // Set to `true` to include a "UDC_IDENTIFIER" attribute with every
  // `/samlValidate` request. The attribute value will be set to the same
  // value as the standard CAS "user" value, i.e. the username. This is to
  // enable compliance with the SIS product Banner 9. Note: this value gets
  // mapped to a value in Banner's `general.gobumap` table.
  saml11Banner9Hack: false,

  // Given a string name the server will attempt to `require` the module.
  // If the string begins with `~/`, it indicates that the "module" is within
  // the server's `lib/plugins/` directory, i.e. is one of the shipped demo
  // plugins.
  //
  // Otherwise, you may specify already required instances.
  plugins: {
    theme: '~/basicTheme',
    attributesResolver: '~/jsAttributesResolver',
    serviceRegistry: '~/jsServiceRegistry',
    ticketRegistry: '~/jsTicketRegistry',
    auth: [
      '~/jsIdP'
    ],
    misc: []
  },

  pluginsConf: {}
}

const config = nixconfig({
  initialConfig,
  parentName: 'jscas',
  parentPath: path.resolve(path.join(__dirname, '..')),
  loaders: Object.assign({}, require('nixconfig-yaml'), require('nixconfig-toml'))
})

module.exports = config
